lib/query.js

```diff
@@ -4007,12 +4013,83 @@ Query.prototype._findAndModify = function(type, callback) {
  */
 
 function _completeOneLean(doc, res, opts, callback) {
+  if (opts.lean && opts.lean.transform) {
+    for (const key of Object.keys(doc)) {
+      if (Array.isArray(doc[key])) {
+        doc[key].forEach((item) => {
+          for (const k in item) {
+            if (typeof item[k] === 'object' && item[k] != null && Object.keys(item[k]).length) {
+              _completeOneLean(item[k], res, opts);
+            }
+            opts.lean.transform({}, item);
+          }
+        });
+      } else if (typeof doc[key] === 'object' && doc[key] != null && Object.keys(doc[key]).length) {
+        _completeOneLean(doc[key], res, opts);
+      }
+    }
+    opts.lean.transform({}, doc);
+    if (callback) {
+      return callback(null, doc);
+    } else {
+      return;
+    }
+  }
   if (opts.rawResult) {
     return callback(null, res);
   }
   return callback(null, doc);
 }
 
+/*!
+ * ignore
+ */
+
+function _completeManyLean(model, docs, fields, userProvidedFields, opts, callback) {
+  if (opts.lean && opts.lean.transform) {
+    if (Array.isArray(docs)) {
+      docs.forEach(doc => {
```

Comment:
I'd really recommend just lifting the logic from https://github.com/mongoosejs/mongoose-lean-getters/blob/master/index.js . That's more tested and handles nuances like subdocument vs nested path and field projections, which this logic doesn't handle. It also avoids adding extra overhead of iterating through the whole document as opposed to just subdocs.